drivers/mfd/stm32_timers.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2019 Max van Kessel
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <soc.h>
+#include <device.h>
+#include "stm32_timers.h"
+
+#include <clock_control/stm32_clock_control.h>
+
+#define DEV_CFG(dev)							\
+	((const struct timer_stm32_config * const)(dev)->config->config_info)
+#define DEV_DATA(dev)							\
+	((struct stm32_timers * const)(dev)->driver_data)
+
+struct timer_stm32_config
+{
+	u32_t tim_base;		    /**< Timer base */
+	struct stm32_pclken pclken; /**< subsystem driving this peripheral */
+
+	u8_t align_mode;
+	u8_t dir;
+	u8_t msm;
+	u8_t slave_mode;
+	u8_t slave_trig;
+	u8_t master_trig;
+	u32_t prescaler;
+};
+
+static inline void __tim_stm32_get_clock(struct device *dev)
+{
+	struct stm32_timers *data = DEV_DATA(dev);
+	struct device *clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+
+	__ASSERT_NO_MSG(clk);
+
+	data->clock = clk;
+}
+
+static u32_t __get_tim_clk(u32_t bus_clk, clock_control_subsys_t *sub_system)
+{
+	struct stm32_pclken *pclken = (struct stm32_pclken *)(sub_system);
+	u32_t tim_clk, apb_psc;
+
+	if (pclken->bus == STM32_CLOCK_BUS_APB1) {
+		apb_psc = CONFIG_CLOCK_STM32_APB1_PRESCALER;
+	}
+#if !defined(CONFIG_SOC_SERIES_STM32F0X) && !defined(CONFIG_SOC_SERIES_STM32G0X)
+	else {
+		apb_psc = CONFIG_CLOCK_STM32_APB2_PRESCALER;
+	}
+#endif
+
+	/*
+	 * If the APB prescaler equals 1, the timer clock frequencies
+	 * are set to the same frequency as that of the APB domain.
+	 * Otherwise, they are set to twice (Ã2) the frequency of the
+	 * APB domain.
+	 */
+	if (apb_psc == 1U) {
+		tim_clk = bus_clk;
+	} else	{
+		tim_clk = bus_clk * 2U;
+	}
+
+	return tim_clk;
+}
+
+static int init(struct device *dev)
+{
+	int err = -EIO;
+	const struct timer_stm32_config *cfg = DEV_CFG(dev);
+	struct stm32_timers *data = DEV_DATA(dev);
+	
+	__tim_stm32_get_clock(dev);
+
+	/* enable clock */
+	if (clock_control_on(data->clock,
+				(clock_control_subsys_t*)&cfg->pclken) == 0) {
+		err = 0;
+	}
+
+	if (err == 0) {
+		TIM_TypeDef *tim = (TIM_TypeDef *)cfg->tim_base;
+		u32_t mode = 0;
+
+		data->tim = tim;
+
+		/* TODO: create binding for me */
+		LL_TIM_SetClockSource(tim, LL_TIM_CLOCKSOURCE_INTERNAL);
+
+		LL_TIM_SetPrescaler(tim, cfg->prescaler - 1);
+
+		/* TODO find a more consistent solution for all soc's */
```

Comment:
Are there stm32 series on which this is not working ?

Reply:
The comment corresponds to `TIM_CR1_DIR_Pos`. Maybe you could give some comments on it.

My opinion: `TIM_CR1_DIR_Pos` is not very portable over several series. It might be different on other devices. (I've seen this on the ADC)

Same goes for: `TIM_CR1_CMS_Pos`, `TIM_SMCR_SMS_Pos`, `TIM_SMCR_TS_Pos`, `TIM_CR2_MMS_Pos`